File: 15/day15.pl
:- ['../lib/io.pl'].

parse([]) --> eos.
parse([S-B|T]) -->
    "Sensor at ", parse_coord(S), ": closest beacon is at ", parse_coord(B), "\n", parse(T).
parse_coord(X/Y) -->
    "x=", integer(X), ", y=", integer(Y).

manhattan(PX/PY, QX/QY, D) :-
    D #= abs(PX - QX) + abs(PY - QY).

% does this sensor block any beacons on line Y, and in which X-range
blockline(Y, (SX/SY)-Beacon, MinX/MaxX) :-
    manhattan(SX/SY, Beacon, D),
    DDY #= abs(SY - Y),
    D #>= DDY,
    DDX #= D - DDY,
    MinX #= SX - DDX, MaxX #= SX + DDX.

simplify_range(PX/PY, QX/QY, PX/QY) :-
    PX #< QX, QX #=< PY, PY #=< QY.
simplify_range(PX/PY, QX/QY, PX/PY) :-
    QX #> PX, QX #< PY, QY #> PX, QY #< PY.

simplify([], []).
simplify([H|T], Simpler) :-
    select(R, T, Rem),
    ( simplify_range(H, R, N) ; simplify_range(R, H, N) ),
    % !, generates same answer multiple times otherwise
    simplify([N|Rem], Simpler).
simplify([H|T], [H|Simpler]) :-
    simplify(T, Simpler).

% each sensor has four edge lines ax+b just beyond its coverage
edges(S-B, Edges) :-
    S = _X/SY,
    blockline(SY, S-B, MinX/MaxX),
    LX #= MinX - 1, % (LX/SY) on edges
    RX #= MaxX + 1, % (RX/SY) on edges
    % rewrite A * LX + B = SY for A = 1 and A = -1
    LB1 #= SY + LX,
    LB2 #= SY - LX,
    RB1 #= SY + RX,
    RB2 #= SY - RX,
    Edges = [1/LB2, 1/RB2, -1/LB1, -1/RB1].

check(C, Sensor-Beacon) :-
    manhattan(Sensor, Beacon, D),
    manhattan(Sensor, C, D2),
    D2 #> D.

run :-
    input_stream(15, parse(Input)),
    convlist(blockline(2000000), Input, Ranges),
    simplify(Ranges, [SX/SY]),
    Ans1 #= abs(SX) + abs(SY),
    write_part1(Ans1),
    % generate 4 edges per diamond just outside sensor range
    maplist(edges, Input, E),
    flatten(E, Edges),
    sort(Edges, Sorted),
    % only include duplicate lines and separate into positive and negative A
    include([C]>>(select(C,Edges,Rem),memberchk(C,Rem)), Sorted, Duplicates),
    include([CX/_]>>(CX#=(-1)), Duplicates, Neg),
    include([CX/_]>>(CX#=1), Duplicates, Pos),
    % find an intersection point
    [X,Y] ins 1..4000000,
    member(A/B, Neg),
    member(C/D, Pos),
    A*X + B #= C*X + D,
    A*X + B #= Y,
    label([X,Y]),
    maplist(check(X/Y), Input),
    Ans2 #= 4000000 * X + Y,
    write_part2(Ans2).

File: 15/day15.go
package main

import (
    "fmt"
    "math"

	"github.com/deosjr/adventofcode2022/lib"
)

type coord struct {
    x, y int64
}

type sensor struct {
    pos coord
    beacon coord
    distance int64
}

func newSensor(p, b coord) sensor {
    d := manhattan(p,b)
    return sensor{pos:p, beacon:b, distance: d}
}

func manhattan(p, q coord) int64 {
    return int64(math.Abs(float64(p.x-q.x))) + int64(math.Abs(float64(p.y-q.y)))
}

func check(sensors []sensor, c coord) bool {
    for _, s := range sensors {
        if c == s.beacon {
            return false
        }
        if manhattan(s.pos, c) <= s.distance {
            return true
        }
    }
    return false
}

func fringe(s sensor) []coord {
    coords := []coord{}
    c := coord{0, s.distance + 1}
    for i:=0; i<=int(s.distance); i++ {
        c = coord{c.x-1, c.y-1}
        coords = append(coords, coord{s.pos.x+c.x, s.pos.y+c.y})
    }
    for i:=0; i<=int(s.distance); i++ {
        c = coord{c.x-1, c.y+1}
        coords = append(coords, coord{s.pos.x+c.x, s.pos.y+c.y})
    }
    for i:=0; i<=int(s.distance); i++ {
        c = coord{c.x+1, c.y+1}
        coords = append(coords, coord{s.pos.x+c.x, s.pos.y+c.y})
    }
    for i:=0; i<=int(s.distance); i++ {
        c = coord{c.x+1, c.y-1}
        coords = append(coords, coord{s.pos.x+c.x, s.pos.y+c.y})
    }
    return coords
}

func day15() {
    sensors := []sensor{}
    lib.ReadFileByLine(15, func(line string) {
        var sx, sy, bx, by int64
        fmt.Sscanf(line, "Sensor at x=%d, y=%d: closest beacon is at x=%d, y=%d", &sx, &sy, &bx, &by)
        sensors = append(sensors, newSensor( coord{sx,sy}, coord{bx,by} ))
    })

    coverage := 0
    million := 1000000
    for x:=-10*million;x<20*million;x++ {
        c := coord{int64(x), int64(2*million)}
        if check(sensors, c) {
            coverage++
        }
    }
    lib.WritePart1("%d", coverage)

Loop:
    for _, s := range sensors {
        for _, c := range fringe(s) {
            if c.x < 0 || c.y < 0 || c.x > 4000000 || c.y > 4000000 {
                continue
            }
            if !check(sensors, c) {
                lib.WritePart2("%d", c.x * 4000000 + c.y)
                break Loop
            }
        }
    }
}

func main() {
    day15()
}
